File: codetraverse/extractors/python_extractor.py
# codetraverse/extractors/python_extractor.py

import os
import json
import chardet
from bs4 import BeautifulSoup
from tree_sitter_language_pack import get_parser
import html
from codetraverse.base.component_extractor import ComponentExtractor

def parse_html_to_text(file_path: str) -> str:
    with open(file_path, 'rb') as f:
        raw = f.read()
    guess = chardet.detect(raw)
    encoding = guess['encoding'] or 'utf-8'
    text = raw.decode(encoding, errors='replace')
    soup = BeautifulSoup(text, 'html.parser')
    plain = html.unescape(soup.get_text(separator='\n'))
    return plain

class PythonComponentExtractor(ComponentExtractor):
    def __init__(self):
        self.parser = get_parser("python")
        self.all_components = []

    def parse_file(self, file_path: str):
        plain = parse_html_to_text(file_path)
        tree = self.parser.parse(plain.encode("utf8"))
        return plain, tree

    def get_text(self, node, plain: str) -> str:
        b = plain.encode("utf8")
        return b[node.start_byte:node.end_byte].decode("utf8", errors="replace")

    def extract_decorators(self, node, plain):
        decs = []
        for c in node.children:
            if c.type == "decorator":
                decs.append(self.get_text(c, plain).strip())
        return decs or None

    def extract_parameters(self, node, plain):
        params = []
        for p in node.named_children:
            if p.type == "identifier":
                params.append({"name": self.get_text(p, plain), "annotation": None})
            elif p.type == "typed_parameter":
                name_node = p.child_by_field_name("name")
                type_node = p.child_by_field_name("type")
                name = self.get_text(name_node, plain) if name_node else None
                ann  = self.get_text(type_node, plain) if type_node else None
                params.append({"name": name, "annotation": ann})
        return params or None

    def extract_type_alias(self, node, plain, module_name):
        name_node  = node.child_by_field_name("name")
        value_node = node.child_by_field_name("value")
        return {
            "kind":       "type_alias",
            "module":     module_name,
            "name":       self.get_text(name_node, plain) if name_node else None,
            "annotation": self.get_text(value_node, plain) if value_node else None,
            "start_line": node.start_point[0] + 1,
            "end_line":   node.end_point[0] + 1,
            "code":       self.get_text(node, plain),
        }

    def extract_lambda(self, node, plain, module_name):
        params = None
        pn = node.child_by_field_name("parameters")
        if pn:
            params = self.extract_parameters(pn, plain)
        return {
            "kind":       "lambda",
            "module":     module_name,
            "name":       "<lambda>",
            "parameters": params,
            "start_line": node.start_point[0] + 1,
            "end_line":   node.end_point[0] + 1,
            "code":       self.get_text(node, plain),
        }

    def extract_yield(self, node, plain, module_name):
        return {
            "kind":       "yield",
            "module":     module_name,
            "expression": self.get_text(node, plain),
            "start_line": node.start_point[0] + 1,
            "end_line":   node.end_point[0] + 1,
        }

    # def extract_comprehension(self, node, plain, module_name):
    #     return {
    #         "kind":       node.type,
    #         "module":     module_name,
    #         "code":       self.get_text(node, plain),
    #         "start_line": node.start_point[0] + 1,
    #         "end_line":   node.end_point[0] + 1,
    #     }

    def extract_function_calls(self, node, plain, module_name, rel_path):
        calls = []

        if node.type == "call":
            # grab the dotted or bare name
            fn = node.child_by_field_name("function") or node.children[0]
            if fn:
                full = self.get_text(fn, plain).strip()
                name = full.split("(")[0]

                # ensure we have a project_root to look under
                pr = getattr(self, "project_root", None)
                if pr is None:
                    pr = os.environ.get("ROOT_DIR") or ""

                resolved = None

                # 1) If it was imported via “from X import Y”
                if name in self.imports:
                    mod = self.imports[name]                        # e.g. "codetraverse.utils.networkx_graph"
                    path = mod.replace(".", "/") + ".py"             # → "codetraverse/utils/networkx_graph.py"
                    # absolute on‐disk path
                    abs_path = os.path.join(pr, path)
                    # ** drop the existence check **, or leave it if you want to prefer bare‐mod
                    # if os.path.isfile(abs_path):
                    resolved = f"{path}::{name}"

                # 2) dotted attribute off an import: X.Y.Z
                elif "." in name:
                    root, *rest = name.split(".")
                    leaf = rest[-1]
                    if root in self.imports:
                        mod = self.imports[root]
                        path = mod.replace(".", "/") + ".py"
                        resolved = f"{path}::{leaf}"

                # 3) fallback to local definition
                if not resolved:
                    resolved = f"{rel_path}::{name}"

                calls.append({
                    "name": name,
                    "base_name": name.split(".")[-1],
                    "resolved_callee": resolved
                })

        # recurse
        for c in node.children:
            calls.extend(self.extract_function_calls(c, plain, module_name, rel_path))
        return calls



    def walk_node(self, node, plain, file_path, root_folder, rel_path):
        """
        Recursively walk the Tree-sitter AST and collect component dicts.
        Skips module-level extraction of any function_definition that lives inside a class.
        """
        comps = []
        module_name = os.path.relpath(file_path, root_folder).replace("\\", "/")

        # helper to know if a node is nested inside any class_definition
        def _inside_class(n):
            p = n.parent
            while p:
                if p.type == "class_definition":
                    return True
                p = p.parent
            return False

        # ————————— imports —————————
        if node.type == "import_statement":
            # e.g. `import os, sys as system`
            code = self.get_text(node, plain).strip()
            parts = code.split(None, 1)
            if len(parts) == 2:
                raw = parts[1]
                # split on commas, handle aliases by keeping the full segment
                for name in raw.split(","):
                    name = name.strip()
                    if not name:
                        continue
                    comps.append({
                        "kind": "import",
                        "module": module_name,
                        "name": name,     # now each imported symbol becomes the .name
                        "code": code,
                        "start_line": node.start_point[0] + 1,
                        "end_line":   node.end_point[0] + 1,
                    })

        if node.type == "import_from_statement":
            # e.g. `from typing import NewType, Union as U`
            code = self.get_text(node, plain).strip()
            tokens = code.split()
            if len(tokens) >= 4 and tokens[0] == "from" and tokens[2] == "import":
                source = tokens[1]
                raw = "".join(tokens[3:])  # join back the imported list
                for name in raw.split(","):
                    name = name.strip()
                    if not name:
                        continue
                    comps.append({
                        "kind": "import",
                        "module": module_name,
                        "name": name,
                        "from": source,
                        "code": code,
                        "start_line": node.start_point[0] + 1,
                        "end_line":   node.end_point[0] + 1,
                    })

        # — assignments (module vars) —
        if node.type == "assignment":
            lhs = node.child_by_field_name("left")
            if lhs and lhs.type == "identifier":
                comps.append({
                    "kind": "variable",
                    "module": module_name,
                    "name": self.get_text(lhs, plain),
                    "code": self.get_text(node, plain),
                    "start_line": node.start_point[0] + 1,
                    "end_line": node.end_point[0] + 1,
                })

        # — type aliases —
        if node.type == "type_alias_statement":
            comps.append(self.extract_type_alias(node, plain, module_name))

        # — only module-level function / async definitions (skip those inside a class) —
        if node.type in ("function_definition", "async_function_definition") and not _inside_class(node):
            name_node = node.child_by_field_name("name")
            name = self.get_text(name_node, plain) if name_node else "<anon>"

            # parameters
            params = None
            pn = node.child_by_field_name("parameters")
            if pn:
                params = self.extract_parameters(pn, plain)

            # return annotation
            returns = None
            rt = node.child_by_field_name("return_type")
            if rt:
                returns = self.get_text(rt, plain)

            # decorators and calls
            decs = self.extract_decorators(node, plain)
            calls = self.extract_function_calls(node, plain, module_name, rel_path)

            comps.append({
                "kind": "async_function" if node.type.startswith("async_") else "function",
                "module": module_name,
                "name": name,
                "decorators": decs,
                "parameters": params,
                "returns": returns,
                "start_line": node.start_point[0] + 1,
                "end_line": node.end_point[0] + 1,
                "code": self.get_text(node, plain),
                "function_calls": calls
            })

        # — class definitions + methods —
        if node.type == "class_definition":
            name_node = node.child_by_field_name("name")
            cls_name = self.get_text(name_node, plain) if name_node else "<anon>"
            decs = self.extract_decorators(node, plain)

            # base-classes in parentheses
            bases = []
            bp = node.child_by_field_name("arguments")
            if bp:
                for b in bp.named_children:
                    bases.append(self.get_text(b, plain))

            comps.append({
                "kind": "class",
                "module": module_name,
                "name": cls_name,
                "decorators": decs,
                "bases": bases or None,
                "start_line": node.start_point[0] + 1,
                "end_line": node.end_point[0] + 1,
                "code": self.get_text(node, plain),
            })

            # now pick up its methods
            body = node.child_by_field_name("body")
            if body:
                for m in body.named_children:
                    if m.type in ("function_definition", "async_function_definition"):
                        mn = m.child_by_field_name("name")
                        mname = self.get_text(mn, plain) if mn else "<anon>"

                        mpn = m.child_by_field_name("parameters")
                        mparams = self.extract_parameters(mpn, plain) if mpn else None

                        mdecs = self.extract_decorators(m, plain)
                        mcalls = self.extract_function_calls(m, plain, module_name, rel_path)

                        comps.append({
                            "kind": "async_method" if m.type.startswith("async_") else "method",
                            "module": module_name,
                            "class": cls_name,
                            "name": mname,
                            "decorators": mdecs,
                            "parameters": mparams,
                            "start_line": m.start_point[0] + 1,
                            "end_line": m.end_point[0] + 1,
                            "code": self.get_text(m, plain),
                            "function_calls": mcalls
                        })

        # # — lambdas —
        # if node.type == "lambda":
        #     comps.append(self.extract_lambda(node, plain, module_name))

        # # — yields —
        # if node.type == "yield":
        #     comps.append(self.extract_yield(node, plain, module_name))

        # # — comprehensions —
        # if node.type in (
        #     "list_comprehension",
        #     "set_comprehension",
        #     "dict_comprehension",
        #     "generator_expression"
        # ):
        #     comps.append(self.extract_comprehension(node, plain, module_name))

        # recurse into children
        for c in node.children:
            comps.extend(self.walk_node(c, plain, file_path, root_folder, rel_path))

        return comps


    def extract_from_file(self, filepath, root_folder, rel_path):
        plain, tree = self.parse_file(filepath)
        return self.walk_node(tree.root_node, plain, filepath, root_folder, rel_path)

    def extract_from_folder(self, folder):
        out = []
        project_root = os.environ.get("ROOT_DIR", "")
        abs_folder = os.path.abspath(folder)
        for root, _, files in os.walk(abs_folder):
            for f in files:
                if f.endswith('.py'):
                    file_path = os.path.join(root, f)
                    rel = os.path.relpath(file_path, project_root).replace(os.sep, "/")
                    out.extend(self.extract_from_file(file_path, abs_folder))
        return out

    def process_file(self, file_path: str):
        plain, tree = self.parse_file(file_path)
        root_folder = os.path.dirname(file_path)

        # build import map as you already do…
        self.imports = {}
        root = tree.root_node
        for child in root.children:
            if child.type == "import_statement":
                code = self.get_text(child, plain).strip()
                parts = code.split()
                if parts[0] == "import":
                    for token in " ".join(parts[1:]).split(","):
                        token = token.strip()
                        if " as " in token:
                            mod, alias = token.split(" as ")
                            self.imports[alias.strip()] = mod.strip()
                        else:
                            self.imports[token.split()[0]] = token.split()[0]
            elif child.type == "import_from_statement":
                code = self.get_text(child, plain).strip()
                toks = code.split()
                if toks[0] == "from" and "import" in toks:
                    src = toks[1]
                    names = code.split("import",1)[1].split(",")
                    for n in names:
                        n = n.strip()
                        if " as " in n:
                            orig, alias = n.split(" as ")
                            # map the alias to the module, not to module.func
                            self.imports[alias.strip()] = src
                        else:
                            self.imports[n] = src

                # make project_root available to extract_function_calls
        project_root = os.environ.get("ROOT_DIR", "")
        self.project_root = project_root

        rel = os.path.relpath(file_path, project_root).replace(os.sep, "/")
        # now extract everything
        raw = self.extract_from_file(file_path, root_folder, rel)

        # *** NEW: stamp each comp with a file_path ***
        project_root = os.environ.get("ROOT_DIR", "")
        for c in raw:
            c["file_path"] = rel
            c.setdefault("module", rel)

        # JSON-filter
        self.all_components = [c for c in raw if self._is_jsonable(c)]


    def _is_jsonable(self, x):
        try:
            json.dumps(x)
            return True
        except Exception:
            return False

    def write_to_file(self, output_path: str):
        os.makedirs(os.path.dirname(output_path), exist_ok=True)
        with open(output_path, "w", encoding="utf8") as f:
            json.dump(self.all_components, f, indent=2, ensure_ascii=False)

    def extract_all_components(self):
        return self.all_components

File: codetraverse/extractors/python_extractor.ts
import * as fs from "fs";
import * as path from "path";
import * as chardet from "chardet";
import * as cheerio from "cheerio";
import Parser, { SyntaxNode } from "tree-sitter";
const Python = require("tree-sitter-python"); // Assuming tree-sitter-python is installed
import { ComponentExtractor } from "../base/component_extractor";
import { Component, FunctionCall, Parameter, Field, Variant, Import, TypeUsed, Literal, Variable, Span } from "../types/types";

// Helper function to parse HTML to plain text
function parseHtmlToText(filePath: string): string {
    const raw = fs.readFileSync(filePath);
    const guess = chardet.detect(raw);
    const encoding = (guess as any).encoding || 'utf-8';
    const text = raw.toString(encoding);
    const $ = cheerio.load(text);
    const plain = $.root().text(); // Use .root() to get the root element and then its text
    return plain;
}

export class PythonComponentExtractor implements ComponentExtractor {
    private parser: Parser;
    private allComponents: Component[] = [];
    private imports: { [key: string]: string } = {}; // To store import mappings
    private projectRoot: string | undefined;

    constructor() {
        this.parser = new Parser();
        this.parser.setLanguage(Python);
    }

    private getText(node: SyntaxNode, plain: string): string {
        return plain.substring(node.startIndex, node.endIndex);
    }

    private extractDecorators(node: SyntaxNode, plain: string): string[] | null {
        const decs: string[] = [];
        for (const child of node.children) {
            if (child.type === "decorator") {
                decs.push(this.getText(child, plain).trim());
            }
        }
        return decs.length > 0 ? decs : null;
    }

    private extractParameters(node: SyntaxNode, plain: string): Parameter[] | null {
        const params: Parameter[] = [];
        for (const p of node.namedChildren) {
            if (p.type === "identifier") {
                params.push({ name: this.getText(p, plain), type: null, filePath: "" });
            } else if (p.type === "typed_parameter") {
                const nameNode = p.childForFieldName("name");
                const typeNode = p.childForFieldName("type");
                const name = nameNode ? this.getText(nameNode, plain) : null;
                const annotation = typeNode ? this.getText(typeNode, plain) : null;
                if (name) {
                    params.push({ name: name, type: annotation, filePath: "" });
                }
            }
        }
        return params.length > 0 ? params : null;
    }

    private extractTypeAlias(node: SyntaxNode, plain: string, moduleName: string): Component {
        const nameNode = node.childForFieldName("name");
        const valueNode = node.childForFieldName("value");
        return {
            kind: "type_alias",
            module: moduleName,
            name: nameNode ? this.getText(nameNode, plain) : null,
            typeSignature: valueNode ? this.getText(valueNode, plain) : null,
            startLine: node.startPosition.row + 1,
            endLine: node.endPosition.row + 1,
            code: this.getText(node, plain),
            filePath: "", // Will be filled later
        };
    }

    private extractFunctionCalls(node: SyntaxNode, plain: string, moduleName: string, relPath: string): FunctionCall[] {
        let calls: FunctionCall[] = [];

        if (node.type === "call") {
            const fn = node.childForFieldName("function") || node.children[0];
            if (fn) {
                const full = this.getText(fn, plain).trim();
                const name = full.split("(")[0];

                let resolved: string | null = null;

                // 1) If it was imported via “from X import Y”
                if (this.imports[name]) {
                    const mod = this.imports[name]; // e.g. "codetraverse.utils.networkx_graph"
                    const modPath = mod.replace(/\./g, "/") + ".py"; // → "codetraverse/utils/networkx_graph.py"
                    resolved = `${modPath}::${name}`;
                }
                // 2) dotted attribute off an import: X.Y.Z
                else if (name.includes(".")) {
                    const parts = name.split(".");
                    const root = parts[0];
                    const leaf = parts[parts.length - 1];
                    if (this.imports[root]) {
                        const mod = this.imports[root];
                        const modPath = mod.replace(/\./g, "/") + ".py";
                        resolved = `${modPath}::${leaf}`;
                    }
                }

                // 3) fallback to local definition
                if (!resolved) {
                    resolved = `${relPath}::${name}`;
                }

                calls.push({
                    name: name,
                    type: "function_call", // Added missing 'type' property
                    base: name.split(".").pop() || name,
                    context: "function_call", // This needs to be more specific based on context
                    modules: [resolved],
                });
            }
        }

        // recurse
        for (const child of node.children) {
            calls = calls.concat(this.extractFunctionCalls(child, plain, moduleName, relPath));
        }
        return calls;
    }

    private walkNode(node: SyntaxNode, plain: string, filePath: string, rootFolder: string, relPath: string): Component[] {
        let comps: Component[] = [];
        const moduleName = path.relative(rootFolder, filePath).replace(/\\/g, "/");

        // helper to know if a node is nested inside any class_definition
        const isInsideClass = (n: SyntaxNode): boolean => {
            let p = n.parent;
            while (p) {
                if (p.type === "class_definition") {
                    return true;
                }
                p = p.parent;
            }
            return false;
        };

        // ————————— imports —————————
        if (node.type === "import_statement") {
            const code = this.getText(node, plain).trim();
            const parts = code.split(/\s+/, 2);
            if (parts.length === 2) {
                const raw = parts[1];
                for (const name of raw.split(",")) {
                    const trimmedName = name.trim();
                    if (!trimmedName) continue;
                    comps.push({
                        kind: "import",
                        module: moduleName,
                        name: trimmedName,
                        code: code,
                        startLine: node.startPosition.row + 1,
                        endLine: node.endPosition.row + 1,
                        filePath: "", // Will be filled later
                    });
                }
            }
        }

        if (node.type === "import_from_statement") {
            const code = this.getText(node, plain).trim();
            const tokens = code.split(/\s+/);
            if (tokens.length >= 4 && tokens[0] === "from" && tokens[2] === "import") {
                const source = tokens[1];
                const raw = code.substring(code.indexOf("import") + "import".length).trim();
                for (const name of raw.split(",")) {
                    const trimmedName = name.trim();
                    if (!trimmedName) continue;
                    comps.push({
                        kind: "import",
                        module: moduleName,
                        name: trimmedName,
                        from: source,
                        code: code,
                        startLine: node.startPosition.row + 1,
                        endLine: node.endPosition.row + 1,
                        filePath: "", // Will be filled later
                    });
                }
            }
        }

        // — assignments (module vars) —
        if (node.type === "assignment") {
            const lhs = node.childForFieldName("left");
            if (lhs && lhs.type === "identifier") {
                comps.push({
                    kind: "variable",
                    module: moduleName,
                    name: this.getText(lhs, plain),
                    code: this.getText(node, plain),
                    startLine: node.startPosition.row + 1,
                    endLine: node.endPosition.row + 1,
                    filePath: "", // Will be filled later
                });
            }
        }

        // — type aliases —
        if (node.type === "type_alias_statement") {
            comps.push(this.extractTypeAlias(node, plain, moduleName));
        }

        // — only module-level function / async definitions (skip those inside a class) —
        if (["function_definition", "async_function_definition"].includes(node.type) && !isInsideClass(node)) {
            const nameNode = node.childForFieldName("name");
            const name = nameNode ? this.getText(nameNode, plain) : "<anon>";

            // parameters
            let params: Parameter[] | null = null;
            const pn = node.childForFieldName("parameters");
            if (pn) {
                params = this.extractParameters(pn, plain);
            }

            // return annotation
            let returns: string | null = null;
            const rt = node.childForFieldName("return_type");
            if (rt) {
                returns = this.getText(rt, plain);
            }

            // decorators and calls
            const decs = this.extractDecorators(node, plain);
            const calls = this.extractFunctionCalls(node, plain, moduleName, relPath);

            comps.push({
                kind: node.type.startsWith("async_") ? "async_function" : "function",
                module: moduleName,
                name: name,
                decorators: decs,
                parameters: params,
                returns: returns,
                startLine: node.startPosition.row + 1,
                endLine: node.endPosition.row + 1,
                code: this.getText(node, plain),
                functionCalls: calls,
                filePath: "", // Will be filled later
            });
        }

        // — class definitions + methods —
        if (node.type === "class_definition") {
            const nameNode = node.childForFieldName("name");
            const clsName = nameNode ? this.getText(nameNode, plain) : "<anon>";
            const decs = this.extractDecorators(node, plain);

            // base-classes in parentheses
            const bases: string[] = [];
            const bp = node.childForFieldName("arguments");
            if (bp) {
                for (const b of bp.namedChildren) {
                    bases.push(this.getText(b, plain));
                }
            }

            comps.push({
                kind: "class",
                module: moduleName,
                name: clsName,
                decorators: decs,
                bases: bases.length > 0 ? bases : null,
                startLine: node.startPosition.row + 1,
                endLine: node.endPosition.row + 1,
                code: this.getText(node, plain),
                filePath: "", // Will be filled later
            });

            // now pick up its methods
            const body = node.childForFieldName("body");
            if (body) {
                for (const m of body.namedChildren) {
                    if (["function_definition", "async_function_definition"].includes(m.type)) {
                        const mn = m.childForFieldName("name");
                        const mname = mn ? this.getText(mn, plain) : "<anon>";

                        const mpn = m.childForFieldName("parameters");
                        const mparams = mpn ? this.extractParameters(mpn, plain) : null;

                        const mdecs = this.extractDecorators(m, plain);
                        const mcalls = this.extractFunctionCalls(m, plain, moduleName, relPath);

                        comps.push({
                            kind: m.type.startsWith("async_") ? "async_method" : "method",
                            module: moduleName,
                            class: clsName,
                            name: mname,
                            decorators: mdecs,
                            parameters: mparams,
                            startLine: m.startPosition.row + 1,
                            endLine: m.endPosition.row + 1,
                            code: this.getText(m, plain),
                            functionCalls: mcalls,
                            filePath: "", // Will be filled later
                        });
                    }
                }
            }
        }

        // recurse into children
        for (const child of node.children) {
            comps = comps.concat(this.walkNode(child, plain, filePath, rootFolder, relPath));
        }

        return comps;
    }

    public parseFile(filePath: string): { plain: string; tree: Parser.Tree } {
        const plain = parseHtmlToText(filePath);
        const options: Parser.Options = {
            bufferSize: 1024 * 1024,
        };
        const tree = this.parser.parse(plain, null, options);
        return { plain, tree };
    }

    public extractFromFile(filepath: string, rootFolder: string, relPath: string): Component[] {
        const { plain, tree } = this.parseFile(filepath);
        return this.walkNode(tree.rootNode, plain, filepath, rootFolder, relPath);
    }

    public extractFromFolder(folder: string): Component[] {
        let out: Component[] = [];
        this.projectRoot = process.env.ROOT_DIR || "";
        const absFolder = path.resolve(folder);

        const files = fs.readdirSync(absFolder, { recursive: true, withFileTypes: true });

        for (const dirent of files) {
            if (dirent.isFile() && dirent.name.endsWith('.py')) {
                const file_path = path.join(dirent.path, dirent.name);
                const rel = path.relative(this.projectRoot, file_path).replace(/\\/g, "/");
                out = out.concat(this.extractFromFile(file_path, absFolder, rel));
            }
        }
        return out;
    }

    public processFile(filePath: string): void {
        const { plain, tree } = this.parseFile(filePath);
        const rootFolder = path.dirname(filePath);

        // build import map
        this.imports = {};
        const root = tree.rootNode;
        for (const child of root.children) {
            if (child.type === "import_statement") {
                const code = this.getText(child, plain).trim();
                const parts = code.split(/\s+/);
                if (parts[0] === "import") {
                    for (const token of parts.slice(1).join(" ").split(",")) {
                        const trimmedToken = token.trim();
                        if (trimmedToken.includes(" as ")) {
                            const [mod, alias] = trimmedToken.split(" as ").map(s => s.trim());
                            this.imports[alias] = mod;
                        } else {
                            this.imports[trimmedToken.split(/\s+/)[0]] = trimmedToken.split(/\s+/)[0];
                        }
                    }
                }
            } else if (child.type === "import_from_statement") {
                const code = this.getText(child, plain).trim();
                const toks = code.split(/\s+/);
                if (toks[0] === "from" && toks.includes("import")) {
                    const srcModule = toks[1];
                    const names = code.substring(code.indexOf("import") + "import".length).split(",");
                    for (const n of names) {
                        const trimmedN = n.trim();
                        if (trimmedN.includes(" as ")) {
                            const [orig, alias] = trimmedN.split(" as ").map(s => s.trim());
                            this.imports[alias] = srcModule; // map the alias to the module
                        } else {
                            this.imports[trimmedN] = srcModule;
                        }
                    }
                }
            }
        }

        this.projectRoot = process.env.ROOT_DIR || "";
        const rel = path.relative(this.projectRoot, filePath).replace(/\\/g, "/");

        // now extract everything
        const raw = this.extractFromFile(filePath, rootFolder, rel);

        // stamp each comp with a file_path
        for (const c of raw) {
            c.filePath = rel;
            c.module = c.module || rel;
        }

        // JSON-filter
        this.allComponents = raw.filter(c => this.isJsonable(c));
    }

    private isJsonable(x: any): boolean {
        try {
            JSON.stringify(x);
            return true;
        } catch (e) {
            return false;
        }
    }

    public writeToFile(outputPath: string): void {
        fs.mkdirSync(path.dirname(outputPath), { recursive: true });
        fs.writeFileSync(outputPath, JSON.stringify(this.allComponents, null, 2), "utf8");
    }

    public extractAllComponents(): Component[] {
        return this.allComponents;
    }
}